src/RestApi/StoreApi/Controllers/CartOrder.php

```diff
@@ -0,0 +1,362 @@
+							$product->get_name()
+						),
+						403
+					);
+				}
+
+				// Queue this reservation for later DB insertion.
+				$stock_to_reserve[ $stocked_product_id ] += $item->get_quantity();
+			}
+
+			$stock_to_reserve = array_filter( $stock_to_reserve );
+
+			if ( $stock_to_reserve ) {
+				$stock_to_reserve_rows = [];
+
+				foreach ( $stock_to_reserve as $product_id => $stock_quantity ) {
+					$stock_to_reserve_rows[] = '(' . esc_sql( $order->get_id() ) . ',"' . esc_sql( $product_id ) . '","' . esc_sql( $stock_quantity ) . '")';
+				}
+
+				$values = implode( ',', $stock_to_reserve_rows );
+
+				$wpdb->query(
+					"REPLACE INTO {$wpdb->wc_reserved_stock} ( order_id, product_id, stock_quantity ) VALUES {$values};" // phpcs:ignore
+				);
+			}
+		} catch ( RestException $e ) {
+			return new RestError( $e->getErrorCode(), $e->getMessage(), array( 'status' => $e->getCode() ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Create order and set props based on global settings.
+	 *
+	 * @param RestRequest $request Full details about the request.
+	 * @return \WC_Order A new order object.
+	 */
+	protected function create_order_from_cart( $request ) {
+		$order = $this->get_order_object();
+		$order->set_status( 'draft' );
+		$order->set_created_via( 'store-api' );
+		$order->set_currency( get_woocommerce_currency() );
+		$order->set_prices_include_tax( 'yes' === get_option( 'woocommerce_prices_include_tax' ) );
+		$order->set_customer_id( get_current_user_id() );
+		$order->set_customer_ip_address( \WC_Geolocation::get_ip_address() );
+		$order->set_customer_user_agent( wc_get_user_agent() );
+
+		$this->set_props_from_cart( $order, $request );
+		$this->create_line_items_from_cart( $order, $request );
+		$this->set_props_from_request( $order, $request );
+
+		$order->save();
+
+		// Store Order ID in session so we can look it up later.
+		WC()->session->set( 'draft_order_id', $order->get_id() );
+
+		return $order;
+	}
+
+	/**
+	 * Get an order object, either using a current draft order, or returning a new one.
+	 *
+	 * @return \WC_Order A new order object.
+	 */
+	protected function get_order_object() {
+		$draft_order_id = WC()->session->get( 'draft_order_id' );
+		$draft_order    = $draft_order_id ? wc_get_order( $draft_order_id ) : false;
+
+		if ( $draft_order && $draft_order->has_status( 'draft' ) && 'store-api' === $draft_order->get_created_via() ) {
+			return $draft_order;
+		}
+
+		return new \WC_Order();
+	}
+
+	/**
+	 * Changes default order status to draft for orders created via this API.
+	 *
+	 * @return string
+	 */
+	public function default_order_status() {
+		return 'draft';
+	}
+
+	/**
+	 * Set order totals based on those calculated by the cart.
+	 *
+	 * @param \WC_Order   $order Object to prepare for the response.
+	 * @param RestRequest $request Full details about the request.
+	 */
+	protected function set_props_from_cart( &$order, $request ) {
```

Comment:
Objects are already passed by reference in PHP so there shouldn't be need for using the `&` here (applies to other places this pattern is used too)?

Also, we should typehint these two arguments.